### PR TITLE
Improve DNS message parsing for mDNS

### PR DESCRIPTION
**Pull request by paxswill**
I also encountered the issues in #878/#765, so decided to take a crack at it. To make it easier to keep track of the return types and arguments, I replaced the data+ptr arguments and tuple return types with giving each function a BytesIO stream and using the position in that stream as the pointer.

As far as standards compliance goes, TXT record parsing was improved so that records longer than 63 bytes could be handled (I think that might've been contributing to #878) and the property keys are now compared case-insensitively to match the RFC better. I also changed some of the encode/decode calls with IDNA (for domain names) and ASCII (as DNS-SD keys are explicitly ASCII only).

I changed `Dns[Header|Question|Resource]` from `collections.namedtuple` to `typing.NamedTuple`, but I'm not completely sure I'm satisfied with them (maybe dataclasses? but that would need the dataclass backport for 3.6. Or an actual class hierarchy for different resource record types?)

With these changes, I'm now able to scan and discover my Apple TV that is on a different subnet (through avahi-reflector), as well as pair with the Apple TV. Sorry about the noise in other files from changing the query type constants into an `IntEnum`.

### CODE REVIEW

**Comment by postlund** [comment]
I have honestly not looked at this yet, I will try to take time later tonight to do so (and test it). But I'm *very* grateful that you have taken time and tried to fix this mess. Truly appreciated!

**Comment by postlund** [comment]
You seem to be on a roll. Just ping me when you are done 😅

File: pyatv/support/mdns.py [hunk 1]
```diff
@@ -22,36 +22,41 @@
 logging.addLevelName(TRAFFIC_LEVEL, "Traffic")
 
 
-class QueryType(enum.IntEnum):
```
Comment (paxswill) [inline pyatv/support/mdns.py]: The diff for this is not great, but `QueryType` and the DNS named tuples were moved further down in the file.

**Comment by paxswill** [comment]
Ugh, those macOS failures. It looks like it's a known issue (pytest-dev/pytest-timeout#8), with a [behavior change in 3.8](https://bugs.python.org/issue38904) making it more visible. I can at least occasionally reproduce it locally, so I'll be poking at this for a bit.

Besides a fix for that (which I suspect is going to be to the testing setup, and probably not in this PR), I'm not planning on making other changes. I tried to keep the commits self-contained, but it does lead to some places being changed a few times (ex: parsing changes, with a later split into different modules).

**Comment by postlund** [comment]
I'm gonna try to review this ASAP, but since it's huge and I only have my phone right now, it will probably take a while 😅

File: pyatv/support/dns.py [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+        encoded_length = len(encoded_label)
+        # Length of the encoded label, in bytes, but a maximum of 63
+        # The maximum is 63 as the upper two bits are used as a flag for name
+        # compression.
+        encoded.append(min(encoded_length, 63))
+        if encoded_length == 0:
+            # If we've reached an empty label, assume this is the last component.
+            # Empty labels (two periods right after each other) isn't legal anyways.
+            break
+        # TODO: Should an error be raised if a label is too long?
```
Comment (postlund) [inline pyatv/support/dns.py:39]: I guess we should not end up here, but maybe log a warning or at least debug about it?

**Review by postlund — CHANGES_REQUESTED** [review]
Doing such a big review thoroughly is really hard, but I've done my best to grasp what you have done and it's a big lift of my mess. So big thank you for that! 😊 Just noted two very minor things otherwise ready to go.

Just gonna throw it out there since you obviously have some domain knowledge... There's another bug (#765) where multicast requests are not reaching other networks when using Protocol-Independent Multicast. Or responses are not revived at least. If that's something you might have any ideas for troubleshooting, I would be very very interested...

File: pyatv/support/dns.py [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+from ipaddress import IPv4Address
+
+from pyatv.support.collections import CaseInsensitiveDict
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def unpack_stream(fmt: str, buffer: typing.BinaryIO) -> typing.Tuple:
+    """Unpack a tuple from a binary stream, much like `struct.unpack`."""
```
[the remark this answers is not in the document]
Comment (paxswill) [reply pyatv/support/dns.py]: I rewrote this docstring to be closer to [`struct.unpack`'s](https://docs.python.org/3/library/struct.html#struct.unpack) (which uses "Unpack from *thing*, according to the format string")

File: pyatv/support/dns.py (hunk 2 above)
Comment (paxswill) [reply pyatv/support/dns.py]: Yeah, I wasn't sure if raising a `ValueError` would be more appropriate than just logging it. If this was part of a more DNS focused-library maybe, but this is just a supporting module for Apple TV support. Speaking of which, I think I remember seeing macOS truncate names when they got too long too (but I haven't had such a long hostname in a while).

**Comment by paxswill** [comment]
> Doing such a big review thoroughly is really hard, but I've done my best to grasp what you have done and it's a big lift of my mess. So big thank you for that! Just noted two very minor things otherwise ready to go.

Yeah, sorry not splitting it up a bit more :sweat_smile:

 In addition to those changes, I also made a small change to `DnsMessage.pack()` to use the `DnsHeader` and `DnsQuestion` `pack()` methods (I thought I'd made that change already :/ ).

> Just gonna throw it out there since you obviously have some domain knowledge... There's another bug (# 765) where multicast requests are not reaching other networks when using Protocol-Independent Multicast. Or responses are not revived at least. If that's something you might have any ideas for troubleshooting, I would be very very interested...

I haven't used Protocol-Independent Multicast before, but I might be able to get a simple setup going to reproduce it after another project I'm poking at.

**Comment by postlund** [comment]
Looks good to me! 👍

**Review by postlund — APPROVED** [review]

**Pull request merged**